arangod/Cluster/ServerState.cpp

```diff
@@ -647,23 +669,30 @@ bool ServerState::registerAtAgencyPhase2(AgencyComm& comm) {
   TRI_ASSERT(!_id.empty() && !_myEndpoint.empty());
 
   while (!application_features::ApplicationServer::isStopping()) {
-    VPackBuilder builder;
-    try {
-      VPackObjectBuilder b(&builder);
-      builder.add("endpoint", VPackValue(_myEndpoint));
-      builder.add("advertisedEndpoint", VPackValue(_advertisedEndpoint));
-      builder.add("host", VPackValue(getHost()));
-      builder.add("version", VPackValue(rest::Version::getNumericServerVersion()));
-      builder.add("versionString", VPackValue(rest::Version::getServerVersion()));
-      builder.add("engine", VPackValue(EngineSelectorFeature::engineName()));
-    } catch (...) {
-      LOG_TOPIC("de625", FATAL, arangodb::Logger::CLUSTER) << "out of memory";
-      FATAL_ERROR_EXIT();
-    }
-
-    auto result = comm.setValue(currentServersRegisteredPref + _id, builder.slice(), 0.0);
+    std::vector<std::shared_ptr<VPackBuilder>> builders;
+
+    auto buildSetOp = [this, &builders](char const* url, VPackValue const& v) {
+      builders.emplace_back(new VPackBuilder);
+      builders.back()->add(v);
+      return AgencyOperation(currentServersRegisteredPref + _id + "/" + url,
+                             AgencyValueOperationType::SET, builders.back()->slice());
```

Comment:
I thought kaveh added a version of `AgencyOperation` where you could pass a `VPackValue`.

```
  template<typename T>
  AgencyOperation(std::string const& key, AgencyValueOperationType opType, T const& value)
    : _key(AgencyCommManager::path(key)), _opType(), _holder(std::make_shared<VPackBuilder>()) {
    _holder->add(VPackValue(value));
    _value = _holder->slice();
    _opType.type = AgencyOperationType::Type::VALUE;
    _opType.value = opType;
  }
```

Reply:
Oh, I did not see that; that makes this whole business more elegant, thanks.

Reply:
For context why I did not change this yet: This code is not available in the devel branch, so I can't use it (yet).